### ISSUE TITLE
Getting different outputs from MLX and Torch mamba

### ISSUE THREAD

**Issue by phy-reac** [alxndrTL/mamba.py#68]
Hi Alex, 

Thank you very much for this wonderful repo, especially for your MLX mamba. I am trying to establish the equivalence of the torch and mlx versions but seem to be getting different answers. Is there a simple explanation? 

```
import torch
from mamba import Mamba, MambaConfig

torch.manual_seed(42)
torch.cuda.manual_seed(42)
batch, length, dim = 2, 64, 16
config = MambaConfig(d_model=16, d_state=16, d_conv=4, expand_factor=2, n_layers=1)

x = torch.linspace(0, 1, batch * length * dim).reshape(batch, length, dim).to("cpu")
model1 = Mamba(config).to("cpu")
y1 = model1(x)

print(y1)

```
gives:

```
tensor([[[-3.6478e-02,  1.6692e-02,  6.6396e-03,  ..., -3.0698e-02,
           6.6586e-05,  1.0190e-03],
         [-3.3583e-02,  5.6842e-03, -1.9530e-02,  ..., -4.3984e-02,
          -3.8652e-03,  2.2307e-02],
         [-5.0080e-02,  2.1328e-02,  1.9177e-04,  ..., -3.2043e-02,
           1.8406e-02,  5.0803e-02],
         ...,
         [ 3.9803e-01,  4.8913e-01,  4.6045e-01,  ...,  4.6657e-01,
           4.7035e-01,  5.1431e-01],
         [ 4.0585e-01,  4.9694e-01,  4.6826e-01,  ...,  4.7439e-01,
           4.7817e-01,  5.2212e-01],
         [ 4.1367e-01,  5.0476e-01,  4.7607e-01,  ...,  4.8221e-01,
           4.8598e-01,  5.2994e-01]],

        [[ 4.5617e-01,  5.0742e-01,  4.8167e-01,  ...,  4.4276e-01,
           4.7690e-01,  4.8769e-01],
         [ 4.7388e-01,  4.8835e-01,  4.6321e-01,  ...,  4.4822e-01,
           4.8100e-01,  5.1176e-01],
         [ 4.4342e-01,  5.2607e-01,  5.0161e-01,  ...,  4.8481e-01,
           5.1545e-01,  5.5307e-01],
         ...,
         [ 8.9839e-01,  9.8933e-01,  9.6057e-01,  ...,  9.6693e-01,
           9.7048e-01,  1.0145e+00],
         [ 9.0621e-01,  9.9715e-01,  9.6839e-01,  ...,  9.7475e-01,
           9.7830e-01,  1.0223e+00],
         [ 9.1403e-01,  1.0050e+00,  9.7620e-01,  ...,  9.8257e-01,
           9.8611e-01,  1.0301e+00]]], grad_fn=<AddBackward0>)
```

And

```
import mlx.core as mx

from mamba_mlx import Mamba, MambaConfig  # Using modified standalone MambaBlock
mx.random.seed(42)
batch, length, dim = 2, 64, 16  
x = mx.linspace(0, 1, batch * length * dim).reshape(batch, length, dim)


config = MambaConfig(
    d_model=dim,
    d_conv=4,
    # expand_factor=2, 
    d_state=16,
    n_layers=1,  
)

model2 = Mamba(config)
y2 = model2(x)

print( y2)
```

```
array([[[4.79025e-09, -2.50009e-08, -2.25647e-09, ..., -4.72123e-09, -8.28904e-09, 2.00269e-08],
        [4.12576e-08, -2.36106e-07, 3.72687e-08, ..., -1.00115e-07, 7.3953e-08, -4.09242e-08],
        [3.13399e-07, -9.72643e-07, -3.98374e-07, ..., -5.45418e-07, 1.84057e-07, 2.57895e-08],
        ...,
        [0.000819054, -0.00102316, -0.000851851, ..., -0.000816714, 0.00019283, 0.0011897],
        [0.00084454, -0.00105771, -0.000881469, ..., -0.000838936, 0.000198948, 0.00122666],
        [0.000870354, -0.00109284, -0.00091163, ..., -0.000861289, 0.000205158, 0.00126408]],
       [[-9.24108e-05, -0.000224389, 0.000243386, ..., -6.29436e-05, 0.000235274, -0.000370316],
        [0.000103894, -0.000650493, -0.00037632, ..., -0.000281652, 0.000180944, -0.000125766],
        [0.000495204, -0.00135857, -0.00117508, ..., -0.00139091, 0.000314959, 0.000785438],
        ...,
        [0.00300574, -0.00439771, -0.00387948, ..., -0.00224942, 0.000787694, 0.00434221],
        [0.00304723, -0.00446854, -0.00394542, ..., -0.00226828, 0.000800775, 0.00440192],
        [0.00308891, -0.0045399, -0.00401192, ..., -0.00228693, 0.000813995, 0.0044619]]], dtype=float32)

```

**Comment by phy-reac** [comment]
Closing this issue as the discrepancy above is entirely due to weights initialisation of the nn.linear layers. However, just fixing the seed for both mamba.py and mamba_mlx does not work but generating weights as by seeding np array and then assigning those to the torch and mlx nn.linear does work. For example something like this for both MLX and torch version will give almost the same outputs to model(x)

```
class MambaBlock(nn.Module):
    def __init__(self, config: MambaConfig):
        super().__init__()

        self.config = config
        torch.manual_seed(42)
        # projects block input from D to 2*ED (two branches)
        # Generate random weights and biases using numpy
        import numpy as np
        np.random.seed(42)
        in_features = config.d_model
        out_features = 2 * config.d_inner
        bound = 1 / math.sqrt(in_features)
        weight_np = np.random.uniform(-bound, bound, (out_features, in_features)).astype(np.float32)
        bias_np = np.random.uniform(-bound, bound, (out_features,)).astype(np.float32)
        self.in_proj = nn.Linear(config.d_model, 2 * config.d_inner, bias=config.bias)
        with torch.no_grad():
            self.in_proj.weight.copy_(torch.from_numpy(weight_np))
            # self.in_proj.bias.copy_(torch.from_numpy(bias_np))
        # PyTorch weight shape: (out_channels, in_channels, kernel_size)
        self.conv1d = nn.Conv1d(in_channels=config.d_inner, out_channels=config.d_inner, 
                              kernel_size=config.d_conv, bias=config.conv_bias, 
                              groups=config.d_inner,
                              padding=config.d_conv - 1)
        
        weight_shape = (config.d_inner, 1, config.d_conv)
        bias_shape = (config.d_inner,)
        
        # Initialize weights using Kaiming uniform
        # bound = 1 / np.sqrt(config.d_conv)  # Adjusted for depthwise conv
        weight_np = np.random.uniform(-0.1, 0.1, weight_shape).astype(np.float32)
        bias_np = np.random.uniform(-0.1, 0.1, bias_shape).astype(np.float32)
        # Assign weights to PyTorch
        with torch.no_grad():
            self.conv1d.weight.copy_(torch.from_numpy(weight_np))
            self.conv1d.bias.copy_(torch.from_numpy(bias_np))
```

There is also a hard-coded config.d_state in mamba_mlx (line 156)

` A = mx.repeat(mx.arange(1., 16 + 1.).reshape([1, 16]), repeats=config.d_inner, axis=0)`

which should be:

`A = mx.repeat(mx.arange(1., config.d_state + 1.).reshape([1, config.d_state]), repeats=config.d_inner, axis=0)`
